p2p/src/peer_manager/peerdb/mod.rs

```diff
@@ -131,11 +131,25 @@ impl<T: NetworkingService> PeerDb<T> {
 
     /// Checks if the given address is already connected.
     pub fn is_address_connected(&self, address: &T::Address) -> bool {
-        self.addresses.contains(address)
+        self.connected_addresses.contains(address)
     }
 
-    pub fn known_addresses(&self) -> Vec<PeerAddress> {
-        self.addresses.iter().map(TransportAddress::as_peer_address).collect()
+    /// Selects requested count of peer addresses from the DB randomly.
+    /// Result could be shared with remote peers over network.
+    pub fn random_known_addresses(&self, count: usize) -> Vec<T::Address> {
+        // TODO: Use something more efficient
```

Comment:
How about this one? https://stackoverflow.com/a/42272866/1317944
Please make sure it works when the number required is less than the one available.

Reply:
Updated, thanks.
We still copy all addresses into a `Vec`, so it's still not very good. But I expect that this part will change, so I think we should not spend too much time on it now.
Requesting more than available should be fine, `choose_multiple` will update the number internally as needed:
```
let amount = ::core::cmp::min(amount, self.len());
```

Reply:
You don't need to copy all addresses. That's the whole point. You should only copy the addresses that are picked randomly by `choose_multiple`. You're currently cloning them twice.

Reply:
Yes, but `choose_multiple` only works with slices and `known_addresses` is a `HashSet`.
I think we can take into account that entries already randomized in `HashSet` and simply do something like this:

```
self.known_addresses.iter().skip(some_random_number).take(count).cloned().collect::<Vec<_>>()
```